makeWrapper: Fail loudly when misused

ahmedtd opened the pull request:
###### Motivation for this change

On Wednesday, I tried to run `hexl-mode` in emacs, and four days later, here I am.

This pull request does two things to `makeWrapper` and `wrapProgram`:
  - First, `makeWrapper` checks that the file it is trying to wrap is a normal, executable file.  If not, it dies with a (sometimes) useful backtrace.  This prevents us from trying to wrap insane things like directories and `.pyc` files.
  - Secondly, `makeWrapper` dies with a backtrace on unsupported arguments.

Additionally, there are some follow-on commits that resolve misuses of `makeWrapper` that I ran into while getting emacs to build.

This pull request resolves #24445 and #3277.

Open question: This pull request will convert a lot of currently-silent breakage into build failures.  Will `nox-review wip` be sufficient to shake all of these issues out?

###### Things done

- [x] Tested using sandboxing
  ([nix.useSandbox](http://nixos.org/nixos/manual/options.html#opt-nix.useSandbox) on NixOS,
    or option `build-use-sandbox` in [`nix.conf`](http://nixos.org/nix/manual/#sec-conf-file)
    on non-NixOS)
- Built on platform(s)
   - [x] NixOS
   - [ ] macOS
   - [ ] Linux
- [ ] Tested compilation of all pkgs that depend on this change using `nix-shell -p nox --run "nox-review wip"`
- [ ] Tested execution of all binary files (usually in `./result/bin/`)
- [ ] Fits [CONTRIBUTING.md](https://github.com/NixOS/nixpkgs/blob/master/.github/CONTRIBUTING.md).

---

ahmedtd commented:
Hmmm, it looks like the set of packages that `nox-review pr 24944` wants to build is a little large for me to be comfortable doing so from my laptop.  If need be, I can spin up a VM somewhere to crank it out.

FRidh commented:
I'm okay with the changes in `pkgs/development/interpreters/python/wrapper.nix`. I have no opinion on the other changes.

Mic92 commented on pkgs/build-support/setup-hooks/wrap-gapps-hook.sh:
```diff
@@ -36,9 +36,12 @@ wrapGAppsHook() {
+    # Loop over files, without falling prey to filenames containing
+    # newlines, backslashes, or other horrible things.
+    find "${prefix}/bin" "${prefix}/libexec" -type f -executable -print0 \
+    | while IFS= read -r -d '' file; do
```
Are you sure `-d ''` is correct here and not `-d '\0'`

Mic92 replied:
Ok. Probably works. I was just reading this [FAQ](http://mywiki.wooledge.org/BashFAQ/020), where variables were set in the loop, which is not the case here.

Mic92 commented on pkgs/development/interpreters/python/wrapper.nix:
```diff
@@ -23,17 +23,29 @@ let
+      for srcstore in "''${srcstores[@]}"; do
+        if [[ -d "$srcstore"/bin ]]; then
+          find "$srcstore"/bin -type f -executable -print0 \
+          | while read -d "" srcfile; do
```
same question here regarding `-d`

Mic92 replied:
The delimiter is probably ok, but I wonder why IFS is not unset here.

ahmedtd replied:
You're correct, according to that link, IFS should be set.

ahmedtd commented:
Hmmm, it looks like nox-review chokes on this PR, because it tries to pull in `lispPackages.hu.dwim.asdf`, which doesn't exist (the correct attribute is `lispPackages."hu.dwim.asdf"`).

The fundamental problem seems to be that nix-env doesn't properly quote that attribute when dumping it with `-qa`.

volth commented:
It could make sense to make ```substituteInPlace``` fail loudly as well (when it found nothing to replace).
Some usages may be needless now or do something unexpected.

ahmedtd commented:
According to [nix issue 1342](https://github.com/NixOS/nix/issues/1342), dots are not allowed in attribute names.  I guess I will open a PR to fix the packages identified there as being problematic.

ahmedtd commented:
@volth, I would be fine adding that to this commit series, but I'm worried that making it affect more packages will make it even harder to merge.

What can I do to help move this PR forward?

volth commented:
Ok, it could be postponed to another PR

peti commented on pkgs/build-support/setup-hooks/make-wrapper.sh:
```diff
@@ -32,26 +36,20 @@ makeWrapper() {
-        fi
-
-        if test "$p" = "--suffix" -o "$p" = "--prefix"; then
+        elif [[ ("$p" == "--suffix") || ("$p" == "--prefix") ]]; then
```
It's very unfortunate that unrelated changes likes this one are mixed into a PR that's supposed to be about `makeWrapper`, really. These changes needlessly increase the size of the patch and make it harder to review, which is going to delay a successful merge (or even prevent it from happening altogether). 😞

ahmedtd replied:
I don't think this increases the size of the rebuild... the PR would include that first hunk (calling `die` for improper targets) regardless, so the rebuild set already includes all packages using `makeWrapper`.

I can split the argument strictness commit into a separate PR, though.

peti replied:
I am not talking about rebuilds. I'm talking about the size of the patch that constitutes this PR i.e. the diff you're suggesting be committed to Nixpkgs.

ahmedtd replied:
Sure, it can be split out

peti commented on pkgs/build-support/setup-hooks/wrap-gapps-hook.sh:
```diff
@@ -36,9 +36,12 @@ wrapGAppsHook() {
-    for i in $prefix/bin/* $prefix/libexec/*; do
-      echo "Wrapping app $i"
-      wrapProgram "$i" "${gappsWrapperArgs[@]}"
+    # Loop over files, without falling prey to filenames containing
```
I don't believe the problem this change attempts to fix actually exists:

~~~
$ touch "this is a test"
$ for n in *this*; do echo "$n"; done
this is a test
~~~

ahmedtd replied:
Before, it was attempting to wrap all files in `$prefix/bin` and `$prefix/libexec`, including non-executable files and directories.  Additionally, it ignored files in subdirectories of these roots.

`emacs`, in particular, was broken because of the attempt to wrap directories.

peti replied:
Well, then maybe the comment is misleading since it suggests the change is intended to fix handling of file names that contain white space? Am I missing something?

ahmedtd replied:
I agree, the comment is misleading.  I will remove it.

peti commented on pkgs/development/interpreters/python/wrapper.nix:
```diff
@@ -23,17 +23,29 @@ let
-              rm -f "$out/bin/$prg"
-              makeWrapper "$path/bin/$prg" "$out/bin/$prg" --set PYTHONHOME "$out"
-            fi
+      # Generate new wrapper scripts for the all the executables found in our
```
I don't understand why this modification is a part of this PR. Is this change *necessary* after the modification of `makeWrapper`? Does it depend on the change of `makeWrapper`? If either of these answers is "no", then this modification should be a separate PR.

ahmedtd replied:
It is necessary --- before, this code would try to wrap anything at top level in `$path/bin`, including non-executable files (in this case, `.pyc` files) and directories.

ahmedtd replied:
Sorry, it wouldn't try to wrap directories, but it would ignore any files in subdirectories of `$path/bin` (strange, but some packages do this).

peti replied:
Could this change conceivably be applied to Nixpkgs without having the change to `makeWrapper` first?

ahmedtd replied:
Yes, it could.  Is that your preferred route?

peti reviewed (changes requested):
I am very much in favor of this modification. I'm not thrilled by the fact that seemingly unrelated changes are mixed up in one PR though.

ahmedtd commented:
OK, everything is separated.  This PR now depends on [https://github.com/NixOS/nixpkgs/pull/25185](PR25185), for the `die` utility function.

peterhoeg commented on pkgs/build-support/setup-hooks/make-wrapper.sh:
```diff
@@ -24,6 +24,10 @@ makeWrapper() {
 
+    # We should only be wrapping executable files.
+    [[ -f "${original}" && -x "${original}" ]] || \
+        die "makeWrapper cannot be invoked on ${prog} because it is not an executable file"
```
Shouldn't that be ```${original}``` instead of ```${prog}``` ?

peti replied:
Please address https://github.com/NixOS/nixpkgs/pull/24944/files#r129479438.

peterhoeg commented on pkgs/build-support/setup-hooks/make-wrapper.sh:
```diff
@@ -131,6 +135,11 @@ filterExisting() {
+    # We should only be wrapping executable files.
+    [[ -f "${prog}" && -x "${prog}" ]] || \
+        die "wrapProgram cannot be invoked on ${prog} because it is not an executable file"
+
```
I would personally prefer moving this into a function to avoid the duplication.

peti reviewed (changes requested):
Once https://github.com/NixOS/nixpkgs/pull/24944#discussion_r129479438 is addressed, this patch is ready for merging (into `staging`) IMHO.

ahmedtd committed:
makeWrapper: Only wrap normal executable files

`makeWrapper` and `wrapProgram` are being invoked on all kinds of
wacky things (usually with the help of bash globs or other machine
assistance).

So far, I have come across `wrapProgram` being invoked on a directory,
as well as on the empty string.

As far as I can tell, it's only valid to invoke these utilities on a
normal (non-directory, non-device) executable file.  This commit
enforces that precondition.

ahmedtd commented:
OK, the patch is rebased on staging and the check has been broken out into its own function, `assertExecutable`.

`nox` is broken right now, but I built a few packages that use `makeWrapper` and they seem to be fine.

peti commented:
@ahmedtd, we cannot merge this patch without https://github.com/NixOS/nixpkgs/pull/25185, and that PR has had merge conflicts and hasn't seen any update since a long time. I would really like to see both PRs merged, like, soon, but I need your help to get that done. If we cannot make progress with these patches in the next couple of days, then I am inclined to declare this a lost cause and to close both PRs.

ahmedtd commented:
Oh, it looks like I missed the activity on the other PR.   I will update it
tonight.

On Aug 7, 2017 12:36 PM, "Peter Simons" <notifications@github.com> wrote:

> @ahmedtd <https://github.com/ahmedtd>, we cannot merge this patch without
> #25185 <https://github.com/NixOS/nixpkgs/pull/25185>, and that PR has had
> merge conflicts and hasn't seen any update since a long time. I would
> really like to see both PRs merged, like, soon, but I need your help to get
> that done. If we cannot make progress with these patches in the next couple
> of days, then I am inclined to declare this a lost cause and to close both
> PRs.
>
> —
> You are receiving this because you were mentioned.
> Reply to this email directly, view it on GitHub
> <https://github.com/NixOS/nixpkgs/pull/24944#issuecomment-320759136>, or mute
> the thread
> <https://github.com/notifications/unsubscribe-auth/AA-FcpM4_t5XQb9E8A9olgfSZKIXs4rAks5sV2czgaJpZM4M-jBF>
> .
>

peti reviewed (changes requested):
I'll merge once that typo pointed out in https://github.com/NixOS/nixpkgs/pull/24944/files#r129479438 is fixed.

ahmedtd commented:
I'm sorry, it looks like I messed up my rebase on to staging.  I had meant to address that in my last push to the branch.

The branch is now properly rebased on staging, and has the check broken out into an `assertExecutable` function.

peti commented on pkgs/build-support/setup-hooks/make-wrapper.sh:
```diff
@@ -1,3 +1,12 @@
+# Assert that FILE exists and is executable
+#
+# assertExecutable FILE
+assertExecutable() {
```
What happens if the given path is a symlink to an executable?

ahmedtd replied:
It is handled properly.  Cases I just tested:

  * non-executable file: Assertion failure
  * symlink to non-executable file: Assertion failure
  * executable file: Success
  * symlink to executable file: Success
  * traversable directory: Assertion failure
  * symlink to traversable directory: Assertion failure

FRidh commented:
`help2man`:
```
Builder called die: Cannot wrap  because it is not an executable file
Backtrace:
7 assertExecutable /nix/store/rp2928gdi9x9n8ragk7804nf4msmhrr8-hook/nix-support/setup-hook
36 makeWrapper /nix/store/rp2928gdi9x9n8ragk7804nf4msmhrr8-hook/nix-support/setup-hook
143 wrapProgram /nix/store/rp2928gdi9x9n8ragk7804nf4msmhrr8-hook/nix-support/setup-hook
73 _callImplicitHook /nix/store/0ifjmsk5s6nqxz5wlkb2c78xf1213dkc-stdenv/setup
86 _eval /nix/store/0ifjmsk5s6nqxz5wlkb2c78xf1213dkc-stdenv/setup
36 runHook /nix/store/0ifjmsk5s6nqxz5wlkb2c78xf1213dkc-stdenv/setup
821 installPhase /nix/store/0ifjmsk5s6nqxz5wlkb2c78xf1213dkc-stdenv/setup
977 genericBuild /nix/store/0ifjmsk5s6nqxz5wlkb2c78xf1213dkc-stdenv/setup
2 main /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh

builder for ‘/nix/store/dy9xw4d9swsj5ynjkw2q2snz5n01hn3v-help2man-1.47.4.drv’ failed with exit code 1
error: build of ‘/nix/store/dy9xw4d9swsj5ynjkw2q2snz5n01hn3v-help2man-1.47.4.drv’ failed

```
https://nix-cache.s3.amazonaws.com/log/dy9xw4d9swsj5ynjkw2q2snz5n01hn3v-help2man-1.47.4.drv

FRidh commented:
The same happens with `wrapPython` that is used by `buildPythonPackage`.
In this case its `bootstrapped-pip` that fails which is essential for building Python packages.
```
patching sources
configuring
no configure script, doing nothing
building
no Makefile, doing nothing
installing

Builder called die: Cannot wrap  because it is not an executable file
Backtrace:
7 assertExecutable /nix/store/rp2928gdi9x9n8ragk7804nf4msmhrr8-hook/nix-support/setup-hook
36 makeWrapper /nix/store/rp2928gdi9x9n8ragk7804nf4msmhrr8-hook/nix-support/setup-hook
143 wrapProgram /nix/store/rp2928gdi9x9n8ragk7804nf4msmhrr8-hook/nix-support/setup-hook
986 genericBuild /nix/store/0ifjmsk5s6nqxz5wlkb2c78xf1213dkc-stdenv/setup
2 main /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh

builder for ‘/nix/store/xv54v5jjdyb4q59my0sikdpy8mkpp2vq-python2.7-bootstrapped-pip-9.0.1.drv’ failed with exit code 1
cannot build derivation ‘/nix/store/6nnj7rishdc06gflj2j4pmi2aq9bx48g-python2.7-pytest-3.0.7.drv’: 1 dependencies couldn't be built
error: build of ‘/nix/store/6nnj7rishdc06gflj2j4pmi2aq9bx48g-python2.7-pytest-3.0.7.drv’ failed
```
The wrapper is defined at https://github.com/NixOS/nixpkgs/blob/master/pkgs/development/interpreters/python/wrap.sh

peti commented:
@ahmedtd, I wonder why those error message don't contain the path that was supposed to be wrapped?

FRidh commented on pkgs/build-support/setup-hooks/make-wrapper.sh:
```diff
@@ -24,6 +33,8 @@ makeWrapper() {
     local params varName value command separator n fileNames
     local argv0 flagsBefore flags
 
+    assertExecutable "${file}"
```
I think this should be `original` instead of `file`

ahmedtd replied:
Oh, how embarrassing.  Sorry.

Mic92 commented:
Dropbox build is affected by this change:

```
Builder called die: Cannot wrap /nix/store/g7fnci9zvasw0jqb6a33f49hx72y58hy-dropbox-32.4.23/opt/dropbox/dropbox because it is not an executable file
Backtrace:
7 assertExecutable /nix/store/4j3i3dggkig9hy6mgf9iwraf1y932s1q-hook/nix-support/setup-hook
36 makeWrapper /nix/store/4j3i3dggkig9hy6mgf9iwraf1y932s1q-hook/nix-support/setup-hook
1001 genericBuild /nix/store/xa5a2jlryx7mwadpxg0zgn83rpf66qxy-stdenv/setup
2 main /nix/store/9krlzvny65gdc8s7kpb6lkx8cd02c25b-default-builder.sh
```
